### PR TITLE
fix(artists): show off-page bulk selection in toolbar (#1227)

### LINKED ISSUE #1227: fix(artists): bulk selection appears lost when sort moves selected artists off-page
## Summary
On `/artists` with multiple pages of results, bulk selection survives sort changes by artist ID (per #1081 fix). However, when a sort moves selected artists to a different page than the current view, there is no on-screen indicator that the selection survived off-page; from the user's perspective the selection appears lost.

## Repro
1. `/artists` with >500 results (2+ pages at default page size)
2. On page 1, bulk-select 3 artists at varied positions
3. Change sort order such that one or more selected artists move to page 2
4. Page 1 now shows zero checkmarks; the 3 are still selected in memory but invisible from this view

## Expected behavior
Pick one and apply consistently:
- (a) Toolbar count includes off-page items, with an off-page indicator like "3 selected (1 on this page, 2 elsewhere)" plus a way to navigate to or filter to selected items, OR
- (b) Selection is intentionally scoped to the current page and clears on page/sort transitions, with that scope behavior documented in the toolbar copy.

## Acceptance criteria
- The user can tell at a glance how many items are selected total and whether any are off-page.
- Cross-page bulk action either works correctly across pages or is gated with explanation when off-page items exist.
- Behavior consistent between list and grid views.

## Notes
Discovered during M46 W4.C UAT for #1081 on 2026-04-26. Out of scope for #1081 (which targets in-page sort persistence specifically).

### PR DESCRIPTION
**Pull request by sydlexius**
## Summary

After #1081 fixed selection persistence across sort, selected artists that the new sort moved to a different page became invisible in the toolbar count, making the selection look lost from the user's view (#1227).

The toolbar now reports the split form "X selected (Y on this page, Z elsewhere)" whenever any selected items are off the current page, with a "Show selected" affordance that filters `/artists` to the current selection via a new `?ids=` query parameter. Pagination propagates `ids=` so paging through a >50-item selection keeps the filter active. When the filter is in effect a dismissible chip on the toolbar exposes "Show all" to clear it.

## Approach

- Server: `ListParams.IDs` + `CountParams.IDs` filter (capped at 1000, empties dropped); `artists.id IN (?, ...)` clause; `?ids=a,b,c` parsed by `parseIDsParam`.
- JS controller: counts visible-checked vs offPage in the same pass that syncs row checkboxes; uses split-form i18n template only when `offPage > 0`.
- New affordance: `#bulk-show-selected` button (hidden until selection > 0); chip + "Show all" rendered server-side when `data.IDs` non-empty.
- Pagination: `PaginationData.IDs` round-trips the filter on next/prev.
- CSS cascade trap avoided: button starts with `hidden` (single display class), JS toggles between `hidden` and `inline-flex`, never co-applied.

## Test plan

- [x] `TestArtistsPage_IDsFilter` (table + grid + exclusion + chip).
- [x] `TestArtistsPage_IDsFilter_Empty` (empty / trailing-comma / whitespace).
- [x] `TestListParams_Validate_*` (caps, drops empties, preserves nil).
- [x] `TestBulkActionBar_OffPageIndicatorWiring` (display-class invariant).
- [x] `TestPaginationData_PreservesIDsThroughPaging`.
- [x] `bash scripts/pre-push-gate.sh` passes; patch coverage 100%.
- [ ] Optional manual: bulk-select across pages, change sort, click "Show selected".

Closes #1227

## Summary by CodeRabbit

* **New Features**
  * Persistent bulk-selection across artist listing pages via an ids query; server and UI support “Show selected”, “Show all”, and a toolbar button that preserves/clears selections and updates history.
  * Selection lists are validated/normalized and capped at 1000 IDs; pagination preserves the selection filter.

* **Tests**
  * Added UI and contract tests covering ids handling, chip/button rendering, pagination preservation, and edge cases.

* **Documentation**
  * Added OpenAPI docs and localized strings for selection UI and status messages.

### CODE REVIEW

File: internal/api/handlers_artist_test.go [hunk 1]
```diff
@@ -212,6 +213,110 @@ func TestArtistsPage_BulkSelectionSurvivesSort(t *testing.T) {
...
+		raw string
+	}{
+		{"absent", ""},
+		{"trailing-comma", ","},
+		{"whitespace-only", " , , "},
+	}
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			url := "/artists?view=table&ids=" + neturl.QueryEscape(tc.raw)
+			req := httptest.NewRequestWithContext(ctx, http.MethodGet, url, nil)
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply internal/api/handlers_artist_test.go:315]: Fixed in 445fcf70. The "absent" subtest now constructs the URL without &ids= so it exercises the raw == "" short-circuit in parseIDsParam.

File: internal/api/handlers_artist.go [hunk 2]
```diff
@@ -110,6 +111,31 @@ func intQuery(r *http.Request, key string, def int) int {
...
+		p = strings.TrimSpace(p)
+		if p != "" {
+			ids = append(ids, p)
+		}
+	}
+	if len(ids) == 0 {
+		return nil
+	}
+	return ids
+}
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply internal/api/handlers_artist.go:141]: Fixed in 445fcf70. parseIDsParam now filters tokens by idPattern (^[A-Za-z0-9_-]{1,64}$, the same canonical shape used by handlers_bulk_actions). Added a malformed-input subtest to pin the rejection.

File: internal/artist/params.go [hunk 3]
```diff
@@ -42,9 +50,16 @@ func (p CountParams) toListParams() ListParams {
...
 		Filters:        p.Filters,
+		IDs:            p.IDs,
 	}
 }
 
+// MaxListIDs caps the number of artist IDs accepted by the IDs filter.
+// Matches the bulk-action cap (handlers_bulk_actions.MaxBulkActionIDs) so
+// the "Show selected" affordance can never request more rows than the
+// in-memory selection store is allowed to hold.
+const MaxListIDs = 1000
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply internal/artist/params.go:62]: Fixed in 445fcf70. MaxBulkActionIDs is now sourced from artist.MaxListIDs so the API request cap and the domain-layer IDs-filter cap can never drift. Direction respects the existing api->artist dep graph.

File: web/templates/artists.templ [hunk 4]
```diff
@@ -1154,6 +1196,29 @@ templ ArtistsPage(assets AssetPaths, data ArtistListData) {
...
+					if (!t || t.id !== 'bulk-show-selected') return;
+					if (selected.size === 0) return;
+					var ids = Array.from(selected).slice(0, MAX_BULK_SELECTION);
+					var params = buildArtistParams({page: '1', ids: ids.join(',')});
+					if (window.htmx) {
+						window.htmx.ajax('GET', _artistsBp + '/artists', {
+							target: '#artist-content',
+							swap: 'outerHTML',
+							values: params
+						});
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply web/templates/artists.templ:1260]: Fixed in 445fcf70. Split showAllURL into showAllPath (root-relative for hx-get) with callers wrapping basePath() for href; updated the bulk-show-selected htmx.ajax call to use the root-relative form so the document-level htmx:configRequest hook prepends basePath exactly once.

**Comment by sydlexius** [comment]
@coderabbitai resolve

File: web/templates/artists_offpage_selection_test.go [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+	// controller wires the click handler to. It also starts hidden so a
+	// fresh page (no selection) does not flash the button on first paint.
+	if !strings.Contains(body, `id="bulk-show-selected"`) {
+		t.Errorf("BulkActionBar missing #bulk-show-selected button; body:\n%s", body)
+	}
+	idx := strings.Index(body, `id="bulk-show-selected"`)
+	// Walk back to the opening <button to confirm it carries the hidden class.
+	btnStart := strings.LastIndex(body[:idx], "<button")
+	if btnStart < 0 {
+		t.Fatalf("could not locate <button opener for #bulk-show-selected")
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply web/templates/artists_offpage_selection_test.go:59]: Fixed in bafdc30f. Switched to t.Fatalf and added an idx < 0 guard so a missing #bulk-show-selected element fails the test deterministically instead of panicking on body[:idx].

File: internal/api/handlers_artist.go [hunk 6]
```diff
@@ -27,6 +27,7 @@ func (r *Router) handleListArtists(w http.ResponseWriter, req *http.Request) {
 		Filter:    req.URL.Query().Get("filter"),
 		LibraryID: req.URL.Query().Get("library_id"),
 		Filters:   parseFlyoutFilters(req),
+		IDs:       parseIDsParam(req.URL.Query().Get("ids")),
 	}
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply internal/api/handlers_artist.go:31]: Fixed in bafdc30f. Documented the ids query parameter under GET /artists in internal/api/openapi.yaml, including the canonical ID-shape filter and the artist.MaxListIDs cap so the spec describes the actual server contract.

File: internal/artist/params_test.go [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+// TestListParams_Validate_PreservesNilIDs documents that the absent-filter
+// case (nil IDs) survives Validate() unchanged. This is the path every
+// existing caller takes today; a regression here would silently shift the
+// list query into "filter to nothing" via an empty IN-clause.
+func TestListParams_Validate_PreservesNilIDs(t *testing.T) {
+	p := ListParams{}
+	p.Validate()
+	if len(p.IDs) != 0 {
+		t.Errorf("Validate populated IDs from nil; got %v", p.IDs)
+	}
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply internal/artist/params_test.go:80]: Fixed in bafdc30f. The assertion now uses != nil so an empty non-nil slice no longer passes; comment explains why nil-vs-empty matters for the SQL builder downstream.

File: web/templates/artists.templ [hunk 8]
```diff
@@ -1279,6 +1369,22 @@ templ ArtistsPage(assets AssetPaths, data ArtistListData) {
...
+					try {
+						var initialIDs = new URLSearchParams(window.location.search).get('ids');
+						if (initialIDs) {
+							var parts = initialIDs.split(',');
+							for (var i = 0; i < parts.length && selected.size < MAX_BULK_SELECTION; i++) {
+								var pid = parts[i].trim();
+								if (pid) selected.add(pid);
+							}
+						}
+					} catch (e) { /* ignore */ }
```
Comment (sydlexius) [inline web/templates/artists.templ]: Fixed in bafdc30f. Client-side hydration now applies the same ^[A-Za-z0-9_-]{1,64}$ regex parseIDsParam uses, so the JS selection store cannot drift from server-validated tokens.

File: internal/artist/params.go [hunk 9]
```diff
@@ -53,6 +69,22 @@ func (p *ListParams) Validate() {
...
+	}
+	if len(p.IDs) > 0 {
+		filtered := p.IDs[:0]
+		for _, id := range p.IDs {
+			if id != "" {
+				filtered = append(filtered, id)
+			}
+		}
+		p.IDs = filtered
+	}
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply internal/artist/params.go:98]: Fixed in bafdc30f. Validate() now drops empties AND deduplicates IDs with first-seen order preserved. Added TestListParams_Validate_DedupesIDs to pin set semantics.

File: web/templates/artists.templ [hunk 10]
```diff
@@ -510,7 +517,12 @@ templ ArtistsPage(assets AssetPaths, data ArtistListData) {
...
 					var urlParams = new URLSearchParams(window.location.search);
 					urlParams.forEach(function(val, key) {
-						if (key.indexOf('filter_') === 0 && !(key in evt.detail.parameters)) {
+						// Carry both filter_* keys (flyout state) and the
+						// ids URL-only param (off-page selection from
+						// "Show selected", #1227) so subsequent search /
+						// sort / view requests do not silently drop the
+						// selection scope and collapse the view.
+						if ((key.indexOf('filter_') === 0 || key === 'ids') && !(key in evt.detail.parameters)) {
 							evt.detail.parameters[key] = val;
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply web/templates/artists.templ]: Fixed in bafdc30f. Reverted the round-1 || key === "ids" addition to the htmx:configRequest listener. Re-injecting ids from window.location.search would have stuffed the stale value into the Show All chip's intentionally ids-less hx-get URL and broken the dismiss flow. The selection scope is still preserved on htmx.ajax flows via buildArtistParams (round-1 change kept) and on pagination via the server-rendered hx-get URLs.

File: web/templates/artists.templ [hunk 11]
```diff
@@ -45,6 +46,12 @@ type ArtistListData struct {
 	LibraryID   string
 	View        string // "table" or "grid"
 	ProfileName string // Active platform profile name for image terminology
+	// IDs reflects the active "Show selected" filter (#1227). When non-empty
+	// the list is restricted to these artist IDs and the toolbar renders a
+	// dismissible chip plus a "Show all" link to clear it. The slice is
+	// preserved across pagination so paging through a large selection
+	// (>1 page) keeps the filter active.
+	IDs []string
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply web/templates/artists.templ:54]: Disagree -- this is a false positive. internal/api/handlers_artist.go:245 sets `IDs: params.IDs` in the ArtistListData struct literal alongside the Pagination.IDs assignment, so data.IDs IS populated and the chip renders correctly. The grep used in the analysis (`data\.IDs\s*=`) does not match the struct-literal field-assignment form (`IDs: params.IDs`), which is why it returned no results. Verified by reading the handler directly.

**Comment by sydlexius** [comment]
@coderabbitai resolve

File: web/templates/artists.templ [hunk 12]
```diff
@@ -644,9 +664,19 @@ templ ArtistsPage(assets AssetPaths, data ArtistListData) {
...
+					// rebuild must mirror whichever the request actually
+					// carried so back-navigation restores the right view.
 					try {
 						var reqUrl = new URL(evt.detail.requestInfo.path, window.location.origin);
+						var ids = reqUrl.searchParams.get('ids');
+						if (ids) {
+							params.ids = ids;
+						} else {
+							delete params.ids;
+						}
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply web/templates/artists.templ:694]: Fixed in 1b194abc. Switched the htmx:afterSwap handler to read evt.detail.requestConfig.path with an evt.detail.xhr.responseURL fallback, mirroring scopeTokenFromRequest so the ids/page reconstruction now uses documented HTMX 2.x fields.

File: web/templates/artists.templ [hunk 13]
```diff
@@ -1279,6 +1371,29 @@ templ ArtistsPage(assets AssetPaths, data ArtistListData) {
...
+						var idRe = /^[A-Za-z0-9_-]{1,64}$/;
+						var initialIDs = new URLSearchParams(window.location.search).get('ids');
+						if (initialIDs) {
+							var parts = initialIDs.split(',');
+							for (var i = 0; i < parts.length && selected.size < MAX_BULK_SELECTION; i++) {
+								var pid = parts[i].trim();
+								if (idRe.test(pid)) selected.add(pid);
+							}
+						}
+					} catch (e) { /* ignore */ }
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply web/templates/artists.templ]: Fixed in 1b194abc. Extracted the seeding logic into seedSelectedFromURL() and now call it from both DOMContentLoaded and the post-scope-clear branch of the inner htmx:afterSwap handler, so search/library/view swaps that retain ?ids= rehydrate the toolbar from the URL instead of leaving it empty next to the server-rendered chip.

File: web/templates/artists.templ [hunk 14]
```diff
@@ -510,6 +517,13 @@ templ ArtistsPage(assets AssetPaths, data ArtistListData) {
...
 					var urlParams = new URLSearchParams(window.location.search);
 					urlParams.forEach(function(val, key) {
+						// Carry only flyout filter_* keys here. The off-page
+						// selection scope (ids, #1227) is carried by
+						// buildArtistParams for htmx.ajax flows and by the
+						// server-rendered hx-get URLs for pagination, so
+						// re-injecting ids from window.location.search would
+						// override the "Show all" chip's intentional
+						// ids-clearing dismiss.
 						if (key.indexOf('filter_') === 0 && !(key in evt.detail.parameters)) {
```
Comment (sydlexius) [inline web/templates/artists.templ:529]: Fixed in 1b194abc. The htmx:configRequest hook now also re-injects ids for declarative hx-get flows (search box, library dropdown, view-toggle), gated on a data-clear-ids="true" opt-out attribute that the "Show all" chip carries so its dismiss path still clears the selection.

**Comment by sydlexius** [comment]
@coderabbitai resolve

**Comment by sydlexius** [comment]
@coderabbitai status

File: web/templates/artists_offpage_selection_test.go [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+				t.Fatalf("chip Show-all link missing hx-get; body:\n%s", body)
+			}
+			hxEnd := strings.Index(body[hxIdx+len(`hx-get="`):], `"`)
+			if hxEnd < 0 {
+				t.Fatalf("malformed hx-get on Show-all link; body:\n%s", body)
+			}
+			hxURL := body[hxIdx+len(`hx-get="`) : hxIdx+len(`hx-get="`)+hxEnd]
+			if strings.Contains(hxURL, "ids=") {
+				t.Errorf("Show-all hx-get must drop ids param; got %q", hxURL)
+			}
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply web/templates/artists_offpage_selection_test.go:166]: Fixed in 0256976e. Added the data-clear-ids="true" assertion alongside the existing aria-label and ids-absent checks so the opt-out attribute is pinned as part of the chip contract, not just the URL shape.

**Comment by sydlexius** [comment]
@coderabbitai resolve

File: web/templates/artists_offpage_selection_test.go [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+		t.Fatalf("render: %v", err)
+	}
+	body := buf.String()
+	if strings.Contains(body, "showing_selected") {
+		t.Errorf("BulkActionBar rendered the selection-filter chip without IDs; body:\n%s", body)
+	}
+	// The "Show all" affordance must only appear on the chip; no chip => no "Show all".
+	if strings.Contains(body, "Show all") {
+		t.Errorf("BulkActionBar rendered Show-all link without IDs filter; body:\n%s", body)
+	}
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply web/templates/artists_offpage_selection_test.go]: Fixed in de99a453. Replaced the "showing_selected" and "Show all" absence checks with a single assertion on the data-clear-ids="true" marker so the test decouples from i18n copy.

File: web/templates/artists_offpage_selection_test.go [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+			hxIdx := strings.Index(body, `hx-get="`)
+			if hxIdx < 0 {
+				t.Fatalf("chip Show-all link missing hx-get; body:\n%s", body)
+			}
+			hxEnd := strings.Index(body[hxIdx+len(`hx-get="`):], `"`)
+			if hxEnd < 0 {
+				t.Fatalf("malformed hx-get on Show-all link; body:\n%s", body)
+			}
+			hxURL := body[hxIdx+len(`hx-get="`) : hxIdx+len(`hx-get="`)+hxEnd]
+			if strings.Contains(hxURL, "ids=") {
```
[the remark this answers is not in the document]
Comment (sydlexius) [reply web/templates/artists_offpage_selection_test.go]: Fixed in de99a453. Anchor the hx-get extraction on the data-clear-ids="true" marker via strings.LastIndex(body[:clearIdx], `hx-get="`) so future hx-get attributes elsewhere in the toolbar cannot bind this assertion to the wrong element.

**Comment by sydlexius** [comment]
@coderabbitai resolve

**Comment by sydlexius** [comment]
@coderabbitai status
